### ISSUE TITLE
Refactor Capture Scalar Outputs to use SymInts/SymFloats

### ISSUE THREAD

**Issue by eellison** [pytorch/torchdynamo#1138]
Currently, in order to support tracking`.item()` calls, `capture_scalar_outputs ` will wrap the output of `.item()` with a 0-element tensor. We should refactor this to work with SymInts/SymFloats for a few reasons:

- fake tensor / proxy tensors will be doing their own internal constant tracking which can distinguish between calls of .item() which are statically known and those that are not, and the torchdynamo `.item()` handling interferes with this 

- this aligns with the other work ongoing to trace out integers/floats operations in the dynamic work stream 

- the current wrapping of scalars to tensors only defers the `.item()` call into the dispatcher. For instance, in adagrad, `param.addcdiv_(grad, std, value=-clr)`, `-clr` in user code is a float but because of `capture_scalar_outputs` in dynamo tracing is a 0-element tensor. when `addcdiv` is invoked in the dispatcher, the `-clr` will get converted to a scalar just as `.item()` call does. See [this issue](https://github.com/pytorch/pytorch/issues/84597) for more context. With [this pr](https://github.com/pytorch/pytorch/pull/83567), this will trigger a graph break, so the current `.item()` handling is insufficient. 
- There might be subtle semantics mismatches in the current handling, e.g., right now we are truncating python floats from double precision to floats.

**Comment by thiagocrepaldi** [comment]
@eellison where can I learn more about "this aligns with the other work ongoing to trace out integers/floats operations in the dynamic work stream? I am working with PyTorch's ONNX converter and properly capturing scalars is a challenge for us too. I've seen fake tensor, proxy tensor and dynamo doing a slightly better work, but if there is a stream of work looking into a definite solution, I am much interested on it :)

**Comment by voznesenskym** [comment]
@eellison do you mind if I do this work on the symbolic-shapes branch?

**Comment by voznesenskym** [comment]
> @eellison where can I learn more about "this aligns with the other work ongoing to trace out integers/floats operations in the dynamic work stream? I am working with PyTorch's ONNX converter and properly capturing scalars is a challenge for us too. I've seen fake tensor, proxy tensor and dynamo doing a slightly better work, but if there is a stream of work looking into a definite solution, I am much interested on it :)

Great question. Our very own @ezyang maintains a really high quality blog here: https://dev-discuss.pytorch.org/t/state-of-symbolic-shapes-branch/777/9 (Linked to latest post)

**Comment by ezyang** [comment]
The main problem with doing this without simultaneously doing dynamic shapes is that the internal rep for symint always builds sympy expressions, but you don't really want without dynamic shapes. I think a coherent design is possible but idk if we feel like doing it right now. Cc @Chillee

**Comment by thiagocrepaldi** [comment]
> The main problem with doing this without simultaneously doing dynamic shapes is that the internal rep for symint always builds sympy expressions, but you don't really want without dynamic shapes. I think a coherent design is possible but idk if we feel like doing it right now. Cc @Chillee

This is something maybe Pytorch ONNX converter team could help with? We are much interested in tackling the dynamic shapes issue for ONNX, maybe we could start there

**Comment by ezyang** [comment]
Actually I think it might not be too hard for ONNX. What you will need to do is implement a SymNode substitute (in the same way XLA has a C++ SymNode impl that is bound to Python) and have it lower to whatever is appropriate for ONNX.

Actually, has ONNX team changed their mind about migrating from TS to make_fx? You might be able to reuse the stock SymNode implementation directly.

**Comment by thiagocrepaldi** [comment]
No, we are still planning on using FX IR instead of TorchScript IR. I will look into SymNode, thanks.
